homeassistant/components/devolo_home_control/devolo_device.py

```diff
@@ -71,7 +72,11 @@ def _sync(self, message: tuple) -> None:
 
     def _generic_message(self, message: tuple) -> None:
         """Handle generic messages."""
-        if len(message) == 3 and message[2] == "battery_level":
+        if (
+            len(message) == 3
+            and message[2] == "battery_level"
+            and self.device_class == SensorDeviceClass.BATTERY
+        ):
```

Comment:
In `_sync`, there's this check:
```py
        if message[0] == self._attr_unique_id:
            self._value = message[1]
        else:
            self._generic_message(message)
```

So as I understand it, `self._generic_message` will be called for any message from the API not related to a `DevoloDeviceEntity`.

The battery sensor was added by PR https://github.com/home-assistant/core/pull/41329, and it seems fairly clear it will not work correctly since a battery level change will overwrite the device's value.
However, will battery sensors work correctly after this change, or will there still be issues with updated for battery sensors overwriting each other?

Reply:
There are two types of messages we can receive from the API. There are messages for changing values which contents of two parts, the uid of a part of the device (pretty similar to enitities in home assistant) 
This are the messages we want to react in the if part in _sync

Then there are the generic messaged. This are messages we want to react party on. In the devolo world they contain information about the whole device. E.g. the battery level, on/offline state.

Battery devices will not overwrite each as a device needs to register itself at the internal publisher. 

https://github.com/home-assistant/core/blob/dev/homeassistant/components/devolo_home_control/devolo_device.py#L54

The first parameter is the device uid. This means that this home assistant entity will only receive messages about the root device. This also means that the heat entity receives the battery mesages and vice versa which causes the bug.

The bug this merge request fixes is the following. Imagine we receive a battery change message. This looks like this:

('hdm:ZWave:F73829FF/2', 100, 'battery_level')

As described earlier, an entity can only subscribe to messages of the whole device (hdm:ZWave:......)
We are the heat entity at the moment
The if expression doesn't match, so generic_messages is called and we match the first if expression in generic_message which overwrites the value of our heat entity.